In [1]:
import string
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold

In [2]:
data = pd.read_csv('2_data.csv')

In [3]:
data

id                                             target  \
0        1  0000100001000100000000001000000000001000010001...   
1        2  0000010001000100010001001000010001000100010010...   
2        3  1000100010000100100001000100010010001000010010...   
3        4  0000010000000100100010000000100001000100010001...   
4        5  0100100000001000100001000000010010001000000001...   
...    ...                                                ...   
1113  1245  1000010000000100010010000000010001000100010001...   
1114  1246  1000000010000100100010000000010000000100010000...   
1115  1247  0100010010001000100001001000010001001000010001...   
1116  1248  1000010000001000100000001000010001001000010010...   
1117  1249  1000100001000100000001000000010010000100100010...   

                                                   teks  
0     wmc AM anemia sign di admiss date report statu...  
1     VH AM discharg summari sign di admiss date rep...  
2     aech AM pnuemonia sign di admiss date report s...  
3     emh AM discharg summari sign di admiss date re...  
4     kfm AM congest heart failur unsign di admiss d...  
...                                                 ...  
1113  chh AM coronari arteri diseas sign di admiss d...  
1114  bcvhc AM discharg summari sign di admiss date ...  
1115  ccvgh AM congest heart failur rule out myocard...  
1116  mmh AM fever unsign di admiss date report stat...  
1117  CH AM asthma sign di admiss date report statu ...  

[1118 rows x 3 columns]

*word_set* adalah himpunan yang menampung seluruh kata dari koleksi dokumen (dokumen latih dan dokumen uji).
<br>Terdapat **24256** kata yang tersebar dalam **1118** kalimat</br>

In [4]:
data_array = data.teks.values
data_array = [x.lower() for x in data_array]
word_set = set()
word_split_list = list()
for x in data_array:
    a = x.split(" ")
    a = [k.replace("'",'') for k in a]
    word_split_list.append(a)
    a = set(a)
    word_set = word_set.union(a)

word_set.remove('')
print("Jumlah Kata: ", len(word_set))
print(word_set)

Jumlah Kata:  24256
{'taha', 'rabal', 'lisinoptil', 'harloff', 'pon', 'andsr', 'bonjorno', 'feet', 'coolbaugh', 'abplanalp', 'christi', 'stofski', 'macul', 'mango', 'cormani', 'amt', 'cuna', 'hypocarb', 'goocharl', 'drumgo', 'skaemp', 'elect', 'rodger', 'monistat', 'misra', 'scmh', 'nmhuhb', 'msec', 'sampl', 'malecki', 'act', 'readdress', 'dho', 'psoriat', 'jerrel', 'mineralocorticoid', 'retic', 'naphthalen', 'retroperitoneum', 'suffic', 'caucasian', 'enbrel', 'suicid', 'prehospit', 'randolph', 'oker', 'mccaster', 'primal', 'hypoparathyroid', 'tmrw', 'shadow', 'bicard', 'abdominojugular', 'worst', 'meridia', 'question', 'cipriani', 'hsil', 'grip', 'suffici', 'bird', 'remondet', 'colonosco', 'grahot', 'nonsurg', 'faintli', 'sirolimu', 'lopressol', 'lomotil', 'trichiura', 'upstair', 'beier', 'helmi', 'rafi', 'fluctuant', 'isidro', 'requip', 'univentricular', 'ruptur', 'aircast', 'tylenol', 'rob', 'distal', 'convuls', 'or', 'tgh', 'zendeja', 'nitrospray', 'pluvios', 'somat', 'rhinorrhea',

Langkah ini menghasilkan matrix sebelum menjadi tf-idf.
<br>Mohon bersabar, langkah ini memakan waktu yang agak lama, sesuai kapasitas komputer yang digunakan.</br>

In [5]:
word_dict_list = list()
for x in data_array:
    word_dict_list.append(dict.fromkeys(word_set, 0))

for x in range (0, len(word_dict_list)):
    wordDict = word_dict_list[x]
    kalimat = word_split_list[x]
    for word in kalimat:
        if(word):
            wordDict[word]+=1
matrix = pd.DataFrame(word_dict_list)
matrix

taha  rabal  lisinoptil  harloff  pon  andsr  bonjorno  feet  coolbaugh  \
0        0      0           0        0    0      0         0     1          0   
1        0      0           0        0    0      0         0     0          0   
2        0      0           0        0    0      0         0     0          0   
3        0      0           0        0    0      0         0     0          0   
4        0      0           0        0    0      0         0     0          0   
...    ...    ...         ...      ...  ...    ...       ...   ...        ...   
1113     0      0           0        0    0      0         1     0          0   
1114     0      0           0        0    0      0         0     0          0   
1115     0      0           0        0    0      0         0     0          0   
1116     0      0           0        0    0      0         0     0          0   
1117     0      0           0        0    0      0         0     0          0   

      abplanalp  ...  miconazol  kareem  disku  skokowski  clemenc  \
0             0  ...          0       0      0          0        0   
1             0  ...          0       0      0          0        0   
2             0  ...          0       0      0          0        0   
3             0  ...          0       0      0          0        0   
4             0  ...          0       0      0          0        0   
...         ...  ...        ...     ...    ...        ...      ...   
1113          0  ...          0       0      0          0        0   
1114          0  ...          0       0      0          0        0   
1115          0  ...          0       0      0          0        0   
1116          0  ...          0       0      0          0        0   
1117          0  ...          0       0      1          0        0   

      determiend  vanderloo  galluss  phosphatas  procardia  
0              0          0        0           0          0  
1              0          0        0           0          0  
2              0          0        0           0          0  
3              0          0        0           0          0  
4              0          0        0           0          0  
...          ...        ...      ...         ...        ...  
1113           0          0        0           0          0  
1114           0          0        0           0          0  
1115           0          0        0           0          0  
1116           0          0        0           1          0  
1117           0          0        0           0          0  

[1118 rows x 24256 columns]

**Memulai Perhitungan TF-IDF**

In [6]:
def computeTF(wordDict, bow):
    tfDict = {}
    bowCount = len(bow)
    for word, count in wordDict.items():
        tfDict[word] = count/float(bowCount)
    return tfDict

def computeIDF(docList):
    import math
    idfDict = {}
    N = len(docList)
    
    idfDict = dict.fromkeys(docList[0].keys(), 0)
    for doc in docList:
        for word, val in doc.items():
            if val > 0:
                idfDict[word] += 1
    
    for word, val in idfDict.items():
        idfDict[word] = math.log10(N / float(val))
        
    return idfDict 

def computeTFIDF(tfBow, idfs):
    tfidf = {}
    for word, val in tfBow.items():
        tfidf[word] = val*idfs[word]
    return tfidf

In [10]:
# TF
tf_list = list()
for x in range(0, len(word_dict_list)):
    tf_list.append(computeTF(word_dict_list[0],word_split_list[0]))
    
# IDF
idfs = computeIDF(word_dict_list)

# TF-IDF
tf_idf_list = list()
for x in tf_list:
    tf_idf_list.append(computeTFIDF(x, idfs))

In [12]:
tfidf = pd.DataFrame(tf_idf_list)

In [13]:
tfidf

taha  rabal  lisinoptil  harloff  pon  andsr  bonjorno      feet  \
0      0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
1      0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
2      0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
3      0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
4      0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
...    ...    ...         ...      ...  ...    ...       ...       ...   
1113   0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
1114   0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
1115   0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
1116   0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   
1117   0.0    0.0         0.0      0.0  0.0    0.0       0.0  0.000955   

      coolbaugh  abplanalp  ...  miconazol  kareem  disku  skokowski  clemenc  \
0           0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
1           0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
2           0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
3           0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
4           0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
...         ...        ...  ...        ...     ...    ...        ...      ...   
1113        0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
1114        0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
1115        0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
1116        0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   
1117        0.0        0.0  ...        0.0     0.0    0.0        0.0      0.0   

      determiend  vanderloo  galluss  phosphatas  procardia  
0            0.0        0.0      0.0         0.0        0.0  
1            0.0        0.0      0.0         0.0        0.0  
2            0.0        0.0      0.0         0.0        0.0  
3            0.0        0.0      0.0         0.0        0.0  
4            0.0        0.0      0.0         0.0        0.0  
...          ...        ...      ...         ...        ...  
1113         0.0        0.0      0.0         0.0        0.0  
1114         0.0        0.0      0.0         0.0        0.0  
1115         0.0        0.0      0.0         0.0        0.0  
1116         0.0        0.0      0.0         0.0        0.0  
1117         0.0        0.0      0.0         0.0        0.0  

[1118 rows x 24256 columns]

Matrix diatas adalah matrix tf-idf. Perhitungan klasifikasi akan dilakukan di notebook selanjutnya.

In [14]:
tfidf.to_csv('3_data.csv', index=False)